### PR TITLE
Search for native artifacts using new layout

### PR DESCRIPTION
**Pull request by cconroy**
Look under META-INF/native/${os.classifier} first to work with a shaded
tcnative.

Blocked on landing and release of companion tcnative PR: https://github.com/netty/netty-tcnative/pull/33

### CODE REVIEW

**Comment by cconroy** [comment]
I'd like to have this fix in 4.x and 3.x. I have a 3.x patch basically ready (that's actually what I tested with since we're currently using). But, this is my first Netty PR: how does the backporting process work? I'm guessing we just do separate backport PRs once this lands?

**Comment by trustin** [comment]
@cconroy Thank you! Let me review soon.

Re: backporting - If your change is applied cleanly to all the branches you want it to be merged, you don't need to send multiple PRs.  However, 3.x and 4x have different directory structure, etc, so you'd better send another PR for 3.x.

File: common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java [hunk 1]
```diff
@@ -144,7 +148,7 @@ private static boolean isWindows() {
     }
 
     private static boolean isOSX() {
-        return OSNAME.startsWith("macosx") || OSNAME.startsWith("osx");
+        return OSNAME.startsWith("osx");
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java]: I just wonder why we included macosx also before. Are you sure we not need this anymore ?

File: common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java [hunk 2]
```diff
@@ -222,6 +225,26 @@ public static void load(String name, ClassLoader loader) {
         }
     }
 
+    private static URL getLibraryUrl(ClassLoader loader, String path) {
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java]: We usually not have get prefixes for this kind of stuff... so maybe just call it `libraryUrl` ?

File: common/src/main/java/io/netty/util/internal/OsDetector.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+    public String getOsName() {
+        return osName;
+    }
+
+    public String getOsArch() {
+        return osArch;
+    }
+
+    public String getOsClassifier() {
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/OsDetector.java]: remove all the get prefixes... We usually not use get prefix in netty

File: common/src/main/java/io/netty/util/internal/OsDetector.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+import io.netty.util.internal.logging.InternalLogLevel;
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+import kr.motd.maven.os.Detector;
+
+
+/**
+ * Use the same OS name and architecture detection and normalization logic as netty/tcnative
+ */
+public class OsDetector extends Detector {
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/OsDetector.java]: final ?

File: common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java (hunk 1 above)
Comment (trustin) [inline common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java]: Some JDK versions used to return 'macosx' although it was rare IIRC.

File: common/src/main/java/io/netty/util/internal/OsDetector.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+package io.netty.util.internal;
+
+import io.netty.util.internal.logging.InternalLogLevel;
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+import kr.motd.maven.os.Detector;
+
+
+/**
+ * Use the same OS name and architecture detection and normalization logic as netty/tcnative
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/OsDetector.java]: Missing `.` on EOL

File: common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java (hunk 1 above)
Comment (cconroy) [inline common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java]: Note: the  OSDetector normalizes osx to always be 'osx'. This is one of the reasons I pulled it in!

File: common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java (hunk 2 above)
Comment (cconroy) [inline common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java]: done

File: common/src/main/java/io/netty/util/internal/OsDetector.java (hunk 3 above)
Comment (cconroy) [inline common/src/main/java/io/netty/util/internal/OsDetector.java]: done

File: common/src/main/java/io/netty/util/internal/OsDetector.java (hunk 4 above)
Comment (cconroy) [inline common/src/main/java/io/netty/util/internal/OsDetector.java]: sure

File: common/src/main/java/io/netty/util/internal/OsDetector.java (hunk 5 above)
Comment (cconroy) [inline common/src/main/java/io/netty/util/internal/OsDetector.java]: fixed

File: common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java (hunk 1 above)
Comment (trustin) [inline common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java]: Oh I see. Missed that one. :-)

File: pom.xml [hunk 6]
```diff
@@ -582,6 +582,12 @@
         <scope>compile</scope>
         <optional>true</optional>
       </dependency>
+      <!-- OS Detection for loading the native libraries -->
+      <dependency>
+        <groupId>kr.motd.maven</groupId>
+        <artifactId>os-maven-plugin</artifactId>
+        <version>1.2.3.Final</version>
+      </dependency>
```
Comment (trustin) [inline pom.xml]: So, we are pulling a new dependency in.  Could we just duplicate the detection code instead of adding a new dependency?

**Comment by trustin** [comment]
@cconroy, it seems like you forgot to push your fixes?

**Comment by cconroy** [comment]
@trustin whoops! I just pushed them.

**Comment by netkins** [comment]
**AUTOMATED MESSAGE for ADMINS**:<br>Please verify and accept the pull request.<br>To accept, say `@netkins accept`<br>To build again, say `@netkins build`<br>To whitelist the author, say `@netkins whitelist`

**Comment by normanmaurer** [comment]
@netkins build

File: pom.xml (hunk 6 above)
Comment (normanmaurer) [inline pom.xml]: @cconroy ping :)

Comment (cconroy) [inline pom.xml]: So sorry: I thought I responded to this, but I guess I figured my response here https://github.com/netty/netty/pull/3502/files#r26598373 covered both questions.

The dependency is very small, and by pulling it in we ensure that both sides are using the same logic to generate the artifact names. I'm not aware of any advantage we gain by copying vs. simply adding this as a dep.

Comment (cconroy) [inline pom.xml]: @trustin if you disagree, i can make your proposed change (or you can feel free to push to this branch if GHE allows that?) Just let me know.

**Comment by cconroy** [comment]
@trustin @normanmaurer ping

**Comment by cconroy** [comment]
@trustin @normanmaurer can we please merge this? once merged, i will prepare companion patches for 3.x and 4.x

**Comment by cconroy** [comment]
@trustin @normanmaurer ping. if you need any small fixups feel free to apply directly and merge.

**Comment by cconroy** [comment]
@normanmaurer why did you close this?

**Comment by normanmaurer** [comment]
hmm... must be a mistake. Sorry :(

**Pull request closed without merging**